pkg/model/diamond.go

```diff
@@ -0,0 +1,257 @@
+)
+
+// IsValid checks the value of a diamond state
+func (s DiamondState) IsValid() bool {
+	switch s {
+	case DiamondInitialized, DiamondRunning, DiamondCommitting, DiamondDone, DiamondCancelling, DiamondCancelled:
+		return true
+	default:
+		return false
+	}
+}
+
+func (s DiamondState) String() string {
+	return string(s)
+}
+
+// SplitState models the running status of an ungoing diamond split
+type SplitState string
+
+const (
+	// SplitDone is the state of a completed split
+	SplitDone SplitState = "done"
+
+	// SplitFailed is the state of a failed split
+	SplitFailed SplitState = "failed"
+
+	// SplitRunning is the state of running split
+	SplitRunning SplitState = "running"
+)
+
+// IsValid checks the value of a diamond state
+func (s SplitState) IsValid() bool {
+	switch s {
+	case SplitDone, SplitFailed, SplitRunning:
+		return true
+	default:
+		return false
+	}
+}
+
+func (s SplitState) String() string {
+	return string(s)
+}
+
+// DiamondDescriptor models a diamond's metadata
+type DiamondDescriptor struct {
+	DiamondID       string            `json:"diamondID" yaml:"diamondID"`
+	StartTime       time.Time         `json:"startTime" yaml:"starTime"` // when the diamond was initialized
+	EndTime         time.Time         `json:"endTime,omitempty" yaml:"endTime,omitempty"`
+	StateChangeTime time.Time         `json:"stateChangeTime" yaml:"stateChangeTime"` // last state change
+	State           DiamondState      `json:"state" yaml:"state"`
+	Mode            ConflictMode      `json:"mode" yaml:"mode"`
+	HasConflicts    bool              `json:"hasConflicts,omitempty" yaml:"hasConflicts,omitempty"`
+	HasCheckpoints  bool              `json:"hasCheckpoints,omitempty" yaml:"hasCheckpoints,omitempty"`
+	Tag             string            `json:"tag,omitempty" yaml:"tag,omitempty"`
+	BundleID        string            `json:"bundleID,omitempty" yaml:"bundleID,omitempty"`
+	Splits          []SplitDescriptor `json:"splits,omitempty" yaml:"splits,omitempty"`
+	_               struct{}
+}
+
+// DiamondDescriptors is a sortable slice of DiamondDescriptor
+type DiamondDescriptors []DiamondDescriptor
+
+func (b DiamondDescriptors) Swap(i, j int) {
+	b[i], b[j] = b[j], b[i]
+}
+func (b DiamondDescriptors) Len() int {
+	return len(b)
+}
+func (b DiamondDescriptors) Less(i, j int) bool {
+	return b[i].DiamondID < b[j].DiamondID
+}
+
+// Last bundle descriptor in slice
+func (b DiamondDescriptors) Last() DiamondDescriptor {
+	return b[len(b)-1]
+}
+
+func defaultDiamondDescriptor() *DiamondDescriptor {
+	ksuid, err := ksuid.NewRandom()
+	if err != nil {
+		panic(fmt.Sprintf("cannot generate random ksuid: %v", err))
+	}
+	return &DiamondDescriptor{
+		DiamondID: ksuid.String(),
+		StartTime: GetBundleTimeStamp(),
+		State:     DiamondInitialized,
+		Mode:      EnableConflicts,
+	}
+}
+
+// NewDiamondDescriptor builds a new DiamondDescriptor
+func NewDiamondDescriptor(opts ...DiamondDescriptorOption) *DiamondDescriptor {
```

Comment:
for consistency [:zombie:  ... err, ⚡️ , i guess?] , i suggest either placing this in `core` or moving `core.NewBDescriptor` to `model`.

consistency has a very early precedence for me by all indications.  i'm considering consistency more important that separation of ideas by patch here.

Reply:
yes this is a refactoring I have already prepared, but will push that PR _after_ diamonds.

Reply:
order of operations considered a nitpick, then.